pkg/client/mount.go

```diff
@@ -258,25 +265,34 @@ func mount(ctxt context.Context, source, dest string, options []string, fs strin
 	dev, err := deviceFromMount(ctxt, source)
 
 	// If we couldn't resolve, then we're probably working with the device already
-	if err != nil || strings.TrimSpace(dev) == "" {
-		dev = source
-	}
-
-	// Check if we're bind-mounting
+	cmd := []string{}
 	bind := false
-	for _, opt := range options {
-		if opt == "--bind" {
-			bind = true
-		}
-	}
 
-	cmd := []string{}
-	// Mount to directory.  If we're bind-mounting we can't specify filesystem
-	if bind {
-		cmd = append([]string{"mount", dev, dest}, options...)
+	if err != nil {
+		dev = source
+		for _, opt := range options {
+			if opt == "--bind" {
+				bind = true
+			}
+		}
+		if bind {
+			cmd = append([]string{"mount", dev, dest}, options...)
+		} else {
+			cmd = append([]string{"mount", "-t", fs, dev, dest}, options...)
+		}
 	} else {
-		cmd = append([]string{"mount", "-t", fs, dev, dest}, options...)
+		// Remove the --bind option from options array
+		for idx, opt := range options {
+			if opt == "--bind" {
+				copy(options[idx:], options[idx+1:])
+				options[len(options)-1] = ""
+				options = options[:len(options)-1]
+				break
+			}
+		}
+		cmd = append([]string{"mount", dev, dest}, options...)
```

Comment:
This might work, but the logic is convoluted.

    if err != nil {
        dev = source # not sure if this is still the right action here, if there is an error, I believe you should error out instead!!!
    }

		for idx, opt := range options {
			if opt == "--bind" {
                                 bind := true # it means we have bind option, so set the flag
                                 copy(options[idx:], options[idx+1:])
				options[len(options)-1] = ""
				options = options[:len(options)-1]
				break
			}
		}

		if bind {
			cmd = append([]string{"mount", dev, dest}, options...)
		} else {
			cmd = append([]string{"mount", "-t", fs, dev, dest}, options...)
		}

This is much more readable IMO.

Reply:
Testing a version of code with returning error instead of assigning 'source' to 'dev'. There are 3 conditions under which deviceFromMount() returns:

(1) When the RunCmd() fails in deviceFromMount(). In this case, mount() will just return back with same error.

(2) When the readlink() fails in deviceFromMount(), it returns the same 'device' that was passed to it. This means mount() already had the correct device to work with (and not a directory). Now in this case, mount() still has 2 different situations. 

(2a) If the options list had --bind option sent, then mount() will remove it since we are already working with correct device and continues with bind operation.

(2b) If the options list didn't have --bind option, then mount() will continue with bind operation with the -t <filesystem_type>.

(3) When deviceFromMount() returns normally, mount() will have the correct device to work with and continue with (2a) and (2b) as described above.

Reply:
If you are confident with your solution and not with the proposed change, keep it.